### PR TITLE
Consolidate the implementation of Type class instances: AssociativeEither - Development

### PR DESCRIPTION
**Pull request by ernestochero**

### CODE REVIEW

**Comment by ernestochero** [comment]
@sideeffffect 
I'm starting with the implementation of #431 
I have a couple of questions and I would like that you respond.
1. Why did you implemented the FutureCommutativeEither using Promise.
2. What means this notation `({ type lambda[+a] = ZIO[R, E, a] })#lambda` (Intellij suggest to me use `ZIO[R, E, +*]`  )
3. Why did  you  use `fa.orElseEither(fb)`  for  AssociativeEither and `fa.raceEither(fb)` for the next one.

Please review this PR before starting adding more code.

Thanks!

**Comment by sideeffffect** [comment]
Thanks for working on this!

> @sideeffffect
> I'm starting with the implementation of #431
> I have a couple of questions and I would like that you respond.
> 
>     1. Why did you implemented the FutureCommutativeEither using Promise.

I don't know, perhaps @adamgfraser would? It was implemented here https://github.com/zio/zio-prelude/pull/174

> 
>     2. What means this notation `({ type lambda[+a] = ZIO[R, E, a] })#lambda` (Intellij suggest to me use `ZIO[R, E, +*]`  )

Dotty can't handle co-/contra-variant placeholders, e.g. `+*` :crying_cat_face: 

> 
>     3. Why did  you  use `fa.orElseEither(fb)`  for  AssociativeEither and `fa.raceEither(fb)` for the next one.

the `either` method should use `orElseEither`. `raceEither` should be used by `eitherPar` (should be done in this ticket https://github.com/zio/zio-prelude/issues/372 ).

> 
> 
> Please review this PR before starting adding more code.
> 
> Thanks!

Please don't forget to run `fix` to make the lint happy.

**Comment by adamgfraser** [comment]
The `CommutativeEither` instance should implement "race" semantics, so we want to return the first `Future` to complete, and I don't believe there is an operator that does this directly for `Future` (`Future.firstCompletedOf` is similar but takes a collection of computations producing the same result type instead of different result types). The `AssociativeEither` instance should implement "orElse" semantics where if the first one fails then we try the second.

We need to do this because the "orElse` interpretation is associative but not commutative. For example, if we do a or else b and both a and b are successful then we would get a back, but if we do b or else a and both are successful we would get b back.

**Comment by ernestochero** [comment]
@sideeffffect 
A question: 
Once we move the commutative instance from commutativeEither to AssociativeEither, Is necessary to keep the associative instance or we need to remove it.  As you mentioned before `It's already implied that they are AssociativeEither just by being CommutativeEither`.

**Comment by ernestochero** [comment]
@sideeffffect  @adamgfraser 
Could you please review this PR?
Thanks!

**Review by sideeffffect — APPROVED** [review]
The PR is fine, thank you :+1:

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 1]
```diff
@@ -156,6 +165,15 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
         fa.map(Left(_)) orElse fb.map(Right(_))
     }
 
+  /**
+   * The `CommutativeEither`  instance for `ZIO`.
+   */
+  implicit def ZIOCommutativeEither[R, E]: CommutativeEither[({ type lambda[+a] = ZIO[R, E, a] })#lambda] =
+    new CommutativeEither[({ type lambda[+a] = ZIO[R, E, a] })#lambda] {
+      def either[A, B](fa: => ZIO[R, E, A], fb: => ZIO[R, E, B]): ZIO[R, E, Either[A, B]] =
+        fa.orElseEither(fb)
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:174]: This should be `raceEither`

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 2]
```diff
@@ -115,6 +115,15 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
...
+   * The `CommutativeEither` instance for `Future`.
+   */
+  implicit def FutureCommutativeEither(implicit ec: ExecutionContext): CommutativeEither[Future] =
+    new CommutativeEither[Future] {
+      def either[A, B](fa: => Future[A], fb: => Future[B]): Future[Either[A, B]] =
+        Promise[Either[A, B]]().completeWith(fa.map(Left(_))).completeWith(fb.map(Right(_))).future
+    }
+
   /**
    * The `AssociativeEither` instance for `Future`.
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:128]: `FutureAssociativeEither` should be removed

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 3]
```diff
@@ -156,6 +165,15 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
...
+   * The `CommutativeEither`  instance for `ZIO`.
+   */
+  implicit def ZIOCommutativeEither[R, E]: CommutativeEither[({ type lambda[+a] = ZIO[R, E, a] })#lambda] =
+    new CommutativeEither[({ type lambda[+a] = ZIO[R, E, a] })#lambda] {
+      def either[A, B](fa: => ZIO[R, E, A], fb: => ZIO[R, E, B]): ZIO[R, E, Either[A, B]] =
+        fa.orElseEither(fb)
+    }
+
   /**
    * The `AssociativeEither` instance for `ZIO`.
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:178]: `ZIOAssociativeEither` should be removed

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 4]
```diff
@@ -222,6 +240,15 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
         fa.map(Left(_)) orElse fb.map(Right(_))
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:240]: `ZStreamAssociativeEither` should be removed

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 5]
```diff
@@ -222,6 +240,15 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
...
+   * The `CommutativeEither` instance for `ZStream`.
+   */
+  implicit def ZStreamCommutativeEither[R, E]: CommutativeEither[({ type lambda[+a] = ZStream[R, E, a] })#lambda] =
+    new CommutativeEither[({ type lambda[+a] = ZStream[R, E, a] })#lambda] {
+      def either[A, B](fa: => ZStream[R, E, A], fb: => ZStream[R, E, B]): ZStream[R, E, Either[A, B]] =
+        fa mergeEither fb
+    }
+
   /**
    * The `AssociativeEither` instance for failed `ZStream`.
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:253]: `ZStreamFailureAssociativeEither` should be removed

**Review by sideeffffect — CHANGES_REQUESTED** [review]
thanks for working on this! There are few things that need to be done :slightly_smiling_face:

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 6]
```diff
@@ -214,28 +212,24 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
...
+  implicit def ZStreamCommutativeEither[R, E]: CommutativeEither[({ type lambda[+a] = ZStream[R, E, a] })#lambda] =
+    new CommutativeEither[({ type lambda[+a] = ZStream[R, E, a] })#lambda] {
       def either[A, B](fa: => ZStream[R, E, A], fb: => ZStream[R, E, B]): ZStream[R, E, Either[A, B]] =
-        fa.map(Left(_)) orElse fb.map(Right(_))
+        fa mergeEither fb
     }
 
   /**
-   * The `AssociativeEither` instance for failed `ZStream`.
+   * The `CommutativeEither` instance for `ZSink`.
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:224]: > (and `AssociativeEither`)

?
:slightly_smiling_face:

Comment (ernestochero) [reply core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:224]: Oh you're right! , let me do the change.

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala (hunk 5 above)
Comment (ernestochero) [reply core/shared/src/main/scala/zio/prelude/AssociativeEither.scala]: @sideeffffect what about with `ZIOFailureAssociativeEither` also should be removed?

Comment (sideeffffect) [reply core/shared/src/main/scala/zio/prelude/AssociativeEither.scala]: `ZIO[R, E, +*]` and `Failure[ZIO[R, +*, A]]` are different types, so both can stay

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 7]
```diff
@@ -214,27 +212,22 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
...
       def either[A, B](fa: => ZStream[R, E, A], fb: => ZStream[R, E, B]): ZStream[R, E, Either[A, B]] =
-        fa.map(Left(_)) orElse fb.map(Right(_))
+        fa mergeEither fb
     }
 
   /**
-   * The `AssociativeEither` instance for failed `ZStream`.
+   * The `CommutativeEither` (and `AssociativeEither`) instance for `ZSink`.
    */
-  implicit def ZStreamFailureAssociativeEither[R, A]
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:228]: We should keep `ZStreamFailureAssociativeEither`, it's about `Failure[ZStream[R, +*, A]]`, so no conflict with `ZStream[R, E, +*]`.

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala [hunk 8]
```diff
@@ -214,27 +212,22 @@ object AssociativeEither extends LawfulF.Invariant[AssociativeEitherDeriveEqualI
...
-    : AssociativeEither[({ type lambda[+e] = Failure[ZStream[R, e, A]] })#lambda] =
-    new AssociativeEither[({ type lambda[+e] = Failure[ZStream[R, e, A]] })#lambda] {
-      def either[EA, EB](
-        fa: => Failure[ZStream[R, EA, A]],
-        fb: => Failure[ZStream[R, EB, A]]
-      ): Failure[ZStream[R, Either[EA, EB], A]] =
-        Failure.wrap {
-          Failure.unwrap(fa).mapError(Left(_)) *> Failure.unwrap(fb).mapError(Right(_))
-        }
+  implicit def ZSinkCommutativeEither[R, E, I, L]
```
Comment (sideeffffect) [inline core/shared/src/main/scala/zio/prelude/AssociativeEither.scala:226]: ZSink goes before ZStream alphabetically

**Review by sideeffffect — CHANGES_REQUESTED** [review]
I've missed few things, my apologies :crying_cat_face:

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala (hunk 8 above)
Comment (ernestochero) [reply core/shared/src/main/scala/zio/prelude/AssociativeEither.scala]: Done!

File: core/shared/src/main/scala/zio/prelude/AssociativeEither.scala (hunk 7 above)
Comment (ernestochero) [reply core/shared/src/main/scala/zio/prelude/AssociativeEither.scala]: Done!

**Comment by ernestochero** [comment]
@sideeffffect don't worry, I like to contribute and I'd like to continue doing that.
Please review and check if I've missed something.

**Review by sideeffffect — APPROVED** [review]
Thanks for contributing! :wink:

**Comment by ernestochero** [comment]
@sideeffffect  
Is there another issue that I could start to work on?  or just could I check the backlog in order to take a new one?

**Comment by sideeffffect** [comment]
Any help on ZIO Prelude would be very much appreciated :pray: 
Have a look around the open issues, the Hackathon ones should be a good pick: https://github.com/zio/zio-prelude/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%23ziohackathon

**Comment by sideeffffect** [comment]
For the record, merging this kinda hinges on how we resolve this question https://github.com/zio/zio-prelude/issues/372#issuecomment-734500103
/cc @adamgfraser

**Comment by sideeffffect** [comment]
As per https://github.com/zio/zio-prelude/issues/372#issuecomment-744067514, we will not be merging this, we decided to take a different path.
Thank you @ernestochero for the nice work you did on this, I'm sorry that it will not get into `master`, but if you would still like to help us with ZIO Prelude, please check our issue tracker, there's still a lot of interesting work to do, especially the [`#ziohackathon`](https://github.com/zio/zio-prelude/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%23ziohackathon) label

**Pull request closed without merging**